pwqgen.1
.\" Copyright (c) 2009 Dmitry V. Levin
.\" All rights reserved.
.\" Copyright (c) 2019 Solar Designer
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd March 10, 2021
.Dt PWQGEN 1
.Os "Openwall Project"
.Sh NAME
.Nm pwqgen
.Nd Generate quality controllable random passphrase
.Sh SYNOPSIS
.Nm Op Ar options
.Sh DESCRIPTION
The
.Nm
program generates a random passphrase using the libpasswdqc library.
Strength of the generated passphrase depends on the amount of randomness
read from
.Pa /dev/urandom .
.Sh OPTIONS
.Bl -tag -width indent
.It Cm random Ns = Ns Ar N
.Pq default: Cm random Ns = Ns 47
The size of randomly-generated passphrase in bits (24 to 136).
.It Cm config Ns = Ns Ar FILE
Load config
.Ar FILE
in the
.Cm passwdqc.conf
format.  This file may define any options described in
.Xr passwdqc.conf 5 , but only the
.Cm random
and
.Cm config
options are honored by
.Nm .
.It Cm --version
Output
.Nm
program version and exit.
.It Cm -h , --help
Output
.Nm
help text and exit.
.El
.Sh EXIT STATUS
.Nm
exits with non-zero status when it encounters invalid config file,
invalid option, invalid parameter value, when it fails to obtain enough
randomness, and in any case when it fails to generate a passphrase.
.Sh FILES
.Pa /etc/passwdqc.conf
(not read unless this suggested file location is specified with the
.Cm config=/etc/passwdqc.conf
option).
.Sh SEE ALSO
.Xr pwqcheck 1 ,
.Xr libpasswdqc 3 ,
.Xr urandom 4 ,
.Xr passwdqc.conf 5 ,
.Xr pam_passwdqc 8 .
.Pp
https://www.openwall.com/passwdqc/
.Sh AUTHORS
The pam_passwdqc module was written for Openwall GNU/*/Linux by Solar Designer.
The
.Nm
program was originally written for ALT GNU/*/Linux by Dmitry V. Levin,
indirectly reusing code from pam_passwdqc (via libpasswdqc).
This manual page was written for Openwall GNU/*/Linux by Dmitry V. Levin.


CHANGES
	Significant changes between 2.0.0 and 2.0.1.

Improved pam_passwdqc's auto-generated policy descriptions, which were
slightly misformatted since the introduction of i18n support in 1.4.0.
Now they not only look prettier, but also make it clearer that the
mentioned lengths are merely the minimums and not the recommended ones.

Updated Russian translation for consistency with the above and to cover
messages added in 1.9.0+.

Increased maximum size of randomly-generated passphrases to 136 bits.
This was already the limit in the underlying API, but the tools' limit
was set to 85.  This increase is to allow for a wider variety of use
cases for the tools.

In the Makefile, use CPPFLAGS and LDFLAGS consistently to be friendlier
to packaging by distros.

Added this file CHANGES based on two latest release announcements, and
started to maintain it.


	Significant changes between 1.9.0 and 2.0.0.

Introduce and use passwdqc_params_free().  This is a minor addition to
the libpasswdqc API related to the addition of external files support.


	Significant changes between 1.5.0 and 1.9.0.

Added support for external wordlist, denylist, and binary filter.  With
these, passwdqc can be configured to deny passwords and passphrases that
are based on lines of a tiny external text file (the "wordlist" option),
directly appear in a tiny external text file (the "denylist" option),
or/and directly appear in a maybe huge binary filter file (the "filter"
option).  While usage of larger external text files is inefficient, the
binary filters are very efficient.

The binary filters can be created and otherwise managed with the newly
added pwqfilter(1) program.  It can create a binary filter from a list
of plaintexts or from MD4 or NTLM hashes.  The latter are supported in a
way that enables importing of HIBP (Pwned Passwords) database revisions
into passwdqc binary filters.  pwqfilter works on arbitrary plain text
strings or hex-encoded hashes, and it can also be reused in lieu of
grep(1) for many purposes, even unrelated to passphrases and security.

Merged changes needed for building with Visual Studio on Windows.  This
includes a refactoring of the random passphrase generator code to make
it shared between platforms.


	Significant changes between 1.4.1 and 1.5.0.

Updated the included wordlist to avoid some inappropriate words in
randomly generated passphrases while not removing any words from the
"word-based" check, and also to have plenty of extra words for
subsequent removal of more words that might be considered inappropriate
from the initial 4096 that are used for randomly generated passphrases.
Most of the added words came from EFF Diceware, BIP-0039, and our own
processing of Project Gutenberg Australia books.


	Significant changes between 1.4.0 and 1.4.1.

Set default for "max" to 72 (was 40).  The previous setting was based on
a reading of RFC 1939, which in practice did not matter.  The new one is
based on bcrypt's truncation at 72, which actually still matters.

Documented "similar" in pwqcheck(1) help message and manual page.  This
is a setting that was supported before and documented for other passwdqc
components before, but was apparently erroneously omitted from here.


	Significant changes between 1.3.2 and 1.4.0.

Implemented i18n support in pam_passwdqc, contributed by Oleg Solovyov,
Andrey Cherepanov, and Dmitry V. Levin.  The i18n support is off by
default, it can be enabled if Linux-PAM is built using --enable-nls
configure option.

Implemented audit support in pam_passwdqc, contributed by Oleg Solovyov
and Dmitry V. Levin.  The audit support is off by default, it can be
enabled if Linux-PAM is built using --enable-audit configure option.

Both of these optional new features had been introduced and are enabled
in ALT Linux distributions, so this version is effectively upstreaming
the changes from there.


	Significant changes between 1.3.1 and 1.3.2.

Compatibility for building with newer versions of glibc, where we now
have to define _DEFAULT_SOURCE for our use of crypt(3).  The problem was
identified and this change tested by Dmitry V. Levin.

Clarified in the man pages that /etc/passwdqc.conf is not read unless
this suggested file location is specified with the config= option.

Clarified the OpenBSD configuration example.

Escape the minus sign in the OpenBSD configuration example to make the
man page linter happy, patch by Jackson Doak via Unit 193.


pwqgen.c
/*
 * Copyright (c) 2008,2009 by Dmitry V. Levin
 * Copyright (c) 2016,2021 by Solar Designer
 * See LICENSE
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "passwdqc.h"

static void
print_help(void)
{
	puts("Generate quality controllable passphrase.\n"
	    "\nUsage: pwqgen [options]\n"
	    "\nValid options are:\n"
	    "  random=N\n"
	    "       set size of randomly-generated passphrase in bits;\n"
	    "  config=FILE\n"
	    "       load config FILE in passwdqc.conf format;\n"
	    "  --version\n"
	    "       print program version and exit;\n"
	    "  -h or --help\n"
	    "       print this help text and exit.");
}

int main(int argc, const char **argv)
{
	passwdqc_params_t params;
	char *reason, *pass;
	int retval;

	if (argc > 1 && argv[1][0] == '-') {
		if (!strcmp("-h", argv[1]) || !strcmp("--help", argv[1])) {
			print_help();
			return 0;
		}

		if (!strcmp("--version", argv[1])) {
			printf("pwqgen version %s\n", PASSWDQC_VERSION);
			return 0;
		}
	}

	passwdqc_params_reset(&params);
	if (argc > 1 &&
	    passwdqc_params_parse(&params, &reason, argc - 1, argv + 1)) {
		fprintf(stderr, "pwqgen: %s\n",
		    (reason ? reason : "Out of memory"));
		free(reason);
		return 1;
	}

	pass = passwdqc_random(&params.qc);
	passwdqc_params_free(&params);
	if (!pass) {
		fprintf(stderr, "pwqgen: Failed to generate a passphrase.\n"
		    "This could happen for a number of reasons: you could have requested\n"
		    "an impossible passphrase length, or the access to kernel random number\n"
		    "pool could have failed.\n");
		return 1;
	}

	setvbuf(stdout, NULL, _IONBF, 0);

	retval = (puts(pass) >= 0 && fflush(stdout) == 0) ? 0 : 1;

	_passwdqc_memzero(pass, strlen(pass));
	free(pass);

	return retval;
}
